Explicitly declare the 'exported' attribute in the AndroidManifest for Android 12

roberto-sartori-gl opened the pull request:
Mandatory starting with Android 12:
https://developer.android.com/about/versions/12/behavior-changes-12#exported

'If your app targets Android 12 or higher and contains activities, services, or broadcast receivers that use intent filters, you must explicitly declare the android:exported attribute for these app components.'

Fixes runtime warning:
'W PackageManager: Failed to parse /vendor/priv-app/TimeKeep: /vendor/priv-app/TimeKeep/TimeKeep.apk (at Binary XML file line #7): com.sony.timekeep.TimeKeep: Targeting S+ (version 31 and above) requires that an explicit value for android:exported be defined when intent filters are present'

Set to 'false' as it was the default in Android versions < 12.

Tested building on an A12 AOSP build system and flashing it on a device (MSM8998). It should be fine also with previous Android versions, but I did not test in those cases.

roberto-sartori-gl committed:
timekeep: explicitly declare the 'exported' attribute in the AndroidManifest

Mandatory starting with Android 12:
https://developer.android.com/about/versions/12/behavior-changes-12#exported

'If your app targets Android 12 or higher and contains activities,
services, or broadcast receivers that use intent filters, you must
explicitly declare the android:exported attribute for these app components.'

Fixes runtime warning:
'W PackageManager: Failed to parse /vendor/priv-app/TimeKeep:
/vendor/priv-app/TimeKeep/TimeKeep.apk (at Binary XML file line #7):
com.sony.timekeep.TimeKeep: Targeting S+ (version 31 and above) requires
that an explicit value for android:exported be defined when intent
filters are present'

Set to 'false' as it was the default in Android versions < 12.

Signed-off-by: Roberto Sartori <roberto.sartori.android@gmail.com>

MarijnS95 commented on AndroidManifest.xml:
```diff
@@ -6,7 +6,8 @@
 
     <application android:label="TimeKeepReceiver">
-        <receiver android:name="TimeKeep">
+        <receiver android:name="TimeKeep"
+        android:exported="false">
```
```suggestion
        <receiver android:name="TimeKeep" android:exported="false">
```

Or:

```suggestion
        <receiver android:name="TimeKeep"
                  android:exported="false">
```

roberto-sartori-gl replied:
Yes, I tested it and it's working fine.
The 'exported' attribute is only used to allow ro deny other applications to explicitly call the activity/service. From Google documentation 

> the activity is accessible to any app, and is launchable by its exact class name.

Intents are always received correctly :)

MarijnS95 replied:
I find [the `<receiver>` documentation](https://developer.android.com/guide/topics/manifest/receiver-element) more clear on this:

> Whether the broadcast receiver can receive messages from **non-system sources** outside its application

(Emphasis mine)

MarijnS95 reviewed (changes requested):
Have you tested this? According to https://developer.android.com/about/versions/12/behavior-changes-12#exported "most cases that do not use the `LAUNCHER` should set this to false", but I wonder if that allows our intent receiver to still receive/subscribe for these actions in the first place - perhaps the system (which delivers broadcasts) is exempt from `exported` in the way that other apps (such as a homescreen/launcher) would otherwise be unable to start this if set to `false`.

MarijnS95 commented:
By the way, indenting log snippets and other code-related blocks with four spaces in your commit message makes them much easier to read in plaintext viewers, and automatically turns it into monospace codeblocks on GitHub (in the rendered commit and the PR description generated from it).

roberto-sartori-gl commented:
> Have you tested this? According to https://developer.android.com/about/versions/12/behavior-changes-12#exported "most cases that do not use the `LAUNCHER` should set this to false", but I wonder if that allows our intent receiver to still receive/subscribe for these actions in the first place - perhaps the system (which delivers broadcasts) is exempt from `exported` in the way that other apps (such as a homescreen/launcher) would otherwise be unable to start this if set to `false`.

Yes, I tested it and it's working fine.
The 'exported' attribute is only used to allow or deny other applications to explicitly call the activity/service. From Google documentation

> the activity is accessible to any app, and is launchable by its exact class name.

Intents are always received correctly :)

MarijnS95 commented:
I find [the `<receiver>` documentation](https://developer.android.com/guide/topics/manifest/receiver-element) more clear on this:

> Whether the broadcast receiver can receive messages from **non-system sources** outside its application

(Emphasis mine)

roberto-sartori-gl commented:
Interesting, reading that it seems that you are correct. The app is still working fine on my device, probably because the messages are sent by the system, like you said.
Do you think we need to set it to true? I don't think we'll need to receive messages from other applications.

MarijnS95 commented:
No it was a question, and as already suggested by the guessed self-answer to that question in the same comment this is probably working because these are system broadcasts. Indeed, that link confirms that is the case.

No action needed on your end besides fixing the indentation, see suggestion above.

roberto-sartori-gl committed:
timekeep: fix indentation in AndroidManifest.xml

Co-authored-by: Marijn Suijten <marijns95@gmail.com>

roberto-sartori-gl committed:
timekeep: remove support for SDK < 28

We are now at SDK 34, no need for this.

It actually causes issues if the /data/time directory is created (for
example, booting one time with permissive SELinux and then switching to
enforcing will cause bootloop).

Signed-off-by: Roberto Sartori <roberto.sartori.android@gmail.com>

roberto-sartori-gl committed:
timekeep: move the Android TimeKeep app to the system partition

Vendor apps cannot access PRIVATE_PLATFORM_APIS anymore.

Signed-off-by: Roberto Sartori <roberto.sartori.android@gmail.com>

jerpelea reviewed (changes requested):
Unfortunately moving the app to the system partition is not an option

roberto-sartori-gl commented:
> Unfortunately moving the app to the system partition is not an option

True, I didn't even remember to have this pull request open. The latest commit was added here automatically by github.

Anyway there will be two options starting (probably) with Android 15:
- the app is moved back to /system
- the app is updated to not use the 'SystemProperties.set()' call

Otherwise, this error does not allow the build to complete:
`error: TimeKeep: Specifies both LOCAL_SDK_VERSION (system_34) and LOCAL_PRIVATE_PLATFORM_APIS (true) but should specify only one`

Without the 'LOCAL_PRIVATE_PLATFORM_APIS', the error is instead:
`TimeKeep.java:68: error: cannot find symbol
                SystemProperties.set(TIMEADJ_PROP, Long.toString(seconds));`

For my personal case, moving the app to system is acceptable. Like I said, it was not my intention to include the commit here :)

edit: actually looking at the Google source code I'm not sure why this error happens only syncing the latest code from the Android main branch. The app still works fine as is on Android 14.

jerpelea commented:
if you have time please update the app API to remove the need for PRIVATE_PLATFORM_APIS

PS. you should not use the master AOSP code

Thanks 
Alin

roberto-sartori-gl commented:
> PS. you should not use the master AOSP code
> 
> Thanks Alin

I use the main AOSP branch for testing, restrictions like these are pushed there way before they are included in a stable tag, but they eventually get there. So it's a way to get ready for new Android releases.

jerpelea commented:
Thanks for the explanation.
It makes sense to use "master".
BR
Alin

MarijnS95 commented:
> Without the 'LOCAL_PRIVATE_PLATFORM_APIS', the error is instead: `TimeKeep.java:68: error: cannot find symbol SystemProperties.set(TIMEADJ_PROP, Long.toString(seconds));`

Looks like it's just using the wrong API to store a _vendor_ property?
